fix(core): call ngOnDestroy on component level services that are provided using a factory function

llorenspujol opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
ngOnDestroy is not called for Services provided via factory function into component level, 

Related Issues: **[28738](https://github.com/angular/angular/issues/28738)**; **[22466](https://github.com/angular/angular/issues/22466#issuecomment-463673658)** (this issue is closed, but it only does it for providers provided at module level, not at component level)

Fixes #28738

## What is the new behavior?
ngOnDestroy is called on services that are provided with a factory function at component level

## Does this PR introduce a breaking change?

- [ ] Yes
- [ x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
This solution is based on the ones that solved the same issue for the factory function services provided at Module level, see: https://github.com/angular/angular/commit/5581e97d2ac6b232cc7402370e681eb909bd63e7

distante commented:
Why was this never reviewed?

Rush commented:
I have the same question, why is this not reviewed?

mhevery reviewed (commented):
LGTM

Could you rebase this on latest master.

JoostK commented on packages/core/test/view/provider_spec.ts:
```diff
@@ -312,6 +311,53 @@ import {ARG_TYPE_VALUES, checkNodeInlineOrDynamic, createRootView, createAndGetR
       });
     });
 
+    describe('destroy', () => {
```
These tests only verify VE behavior, not Ivy. Support for this setup should probably also be tested using an acceptance test (e.g. in `packages/core/test/acceptance/di_spec.ts`), as those run against both renderers. I believe you'll find the Ivy behavior to be different, given that this fix only concerns VE at the moment and Ivy also doesn't seem to call `ngOnDestroy` for component-level providers.

JoostK commented on packages/core/src/view/provider.ts:
```diff
@@ -251,8 +251,20 @@ function _createProviderInstance(view: ViewData, def: NodeDef): any {
+        def.flags |= NodeFlags.OnDestroy;
+        view.def.nodeFlags |= NodeFlags.OnDestroy;
+        if (def.parent) {
+          def.parent.childFlags |= NodeFlags.OnDestroy;
```
Only updating the flags of the direct parent sounds insufficient to me, as I believe that `childFlags` is transitive starting from the root node.

JoostK replied:
Additionally, once `NodeFlags.OnDestroy` is set then VE will unconditionally call the `ngOnDestroy` method:

https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/core/src/view/provider.ts#L575-L577

This introduces a problem, as `def` is shared across all component instances but different instances could return different service implementations from their `useFactory` function, so having the `NodeFlags.OnDestroy` flag set on the shared definition does no longer guarantees that an `ngOnDestroy` method is actually present on those different service implementations.

llorenspujol replied:
@JoostK Can you write some use case where these issues happens? for example a test case, or just a draft... I am unable to reproduce both possible problems that you describe.

For example I am debugging all the process and I don't see childFlags transitive from the root node, I may miss something in the process... So, a use case where this issue happens would be very useful!

Thanks in advance!

JoostK replied:
@llorenspujol Thanks for looking into it. I'll try provide you with a test later today/tomorrow.

JoostK replied:
@llorenspujol Here's the scenario's that would fail in VE but work in Ivy: https://github.com/JoostK/angular/commit/2d6cb9bb3bcb9d4d3dea9dcac764cb92a41defe9. I patched the VE implementation to make it work.

llorenspujol replied:
Thanks for the reproductions, you really nailed both:), I was miss understanding the 'childs' meaning a little bit:), now it's clear.

mhevery reviewed (changes requested):
Those are excellent points @JoostK! Yes at a minimum we should move the tests into acceptance folder to verify that same behavior is in Ivy.

mhevery commented:
This PR was discussed in a team meeting and here are the conclusions:
- We should treat issues which affect ViewEngine in the same way as LTS. Only regressions.
- If the issue affects both ViewEngine and Ivy, than we should fix both so that we can make migration easier. 

So for this PR it has to be refactored by moving the test to acceptance `packages/core/test/acceptance/di_spec.ts` so that we can ensure that the behavior is consistent between ViewEngine and Ivy.

JoostK commented on packages/core/test/acceptance/providers_spec.ts:
```diff
@@ -508,6 +508,75 @@ describe('providers', () => {
           expect(destroyCalls).toBe(0);
         });
 
+    it('should invoke ngOnDestroy for useFactory providers provided in Component', () => {
```
Consider adding tests for factories that return `null`/`undefined`, and a factory the returns multiple different service types (where not all of the services implement the `ngOnDestroy` hook)

JoostK commented on packages/core/src/render3/di_setup.ts:
```diff
@@ -179,6 +179,17 @@ function resolveProvider(
+ * present.
+ */
+function ngOnDestroyForFactoryProviders(this: any) {
+  if (this && this.ngOnDestroy) {
+    this.ngOnDestroy();
+  }
```
As `this.ngOnDestroy` is a megamorphic read, consider reading it only once:

```suggestion
  const onDestroy = this && this.ngOnDestroy;
  if (onDestroy) {
    onDestroy.call(this);
  }
```

llorenspujol replied:
Can you provide me some performance tests?

just kidding lol, thanks for pointing it out:)

JoostK replied:
😅

JoostK commented on packages/core/test/acceptance/providers_spec.ts:
```diff
@@ -508,6 +508,227 @@ describe('providers', () => {
+      expect(logs).toEqual(['OnDestroy']);
+    });
+
+    it('should invoke ngOnDestroy for useFactory providers and account for different service implementations',
```
Can you duplicate this test and replace the `InjectableWithoutDestroyHook` return value in the factory with `null`, such that we verify that we don't read `ngOnDestroy` from the null reference (that's why we do need the `InjectableWithDestroyHook` in order to register the `ngOnDestroy` hook in the first place).

JoostK commented on packages/core/src/view/provider.ts:
```diff
@@ -256,8 +256,24 @@ function _createProviderInstance(view: ViewData, def: NodeDef): any {
+        view.def.nodeFlags |= NodeFlags.OnDestroy;
+        if (def.parent) {
+          let parent: NodeDef|null = def.parent;
+          while (parent !== null) {
```
I just realized we may exit the loop early if possible:

```suggestion
          while (parent !== null && !(parent.childFlags & NodeFlags)) {
```

llorenspujol replied:
well seen!

llorenspujol commented on packages/core/src/view/provider.ts:
```diff
@@ -256,8 +256,24 @@ function _createProviderInstance(view: ViewData, def: NodeDef): any {
+        def.flags |= NodeFlags.OnDestroy;
+        view.def.nodeFlags |= NodeFlags.OnDestroy;
+        if (def.parent) {
+          let parent: NodeDef|null = def.parent;
```
done!

JoostK commented:
@googlebot I consent.

JoostK commented:
@llorenspujol FYI, I rebased on latest master and pushed a fixup commit to get CI green.

AndrewKushnir commented:
[Presubmit (Ivy)](http://test/OCL:334684689:BASE:334684710:1601502775540:5f9b22f7), [Presubmit (VE)](http://test/OCL:334684689:BASE:334684710:1601502775492:1edcf330).

AndrewKushnir commented:
Quick update after running tests in Google's codebase: it looks like this change breaks a number of targets. The good thing is that the errors look similar in ViewEngine and Ivy. The error is somehow related to Forms package as well, so we'll need to investigate the root cause.

llorenspujol commented:
@AndrewKushnir Can you provide a test case or a full reproduction (stackblitz) where this happens?

AndrewKushnir commented:
Hi @llorenspujol,

I performed initial investigation and found out that the issue happens in the "hierarchical forms" use-case, where `ControlContainer` token is initialized via `useFactory` to get a hold of a parent Form, created in a different component. Here is a [Stackblitz example](https://stackblitz.com/edit/angular-ivy-aqhgda?file=src%2Fapp%2Fapp.component.ts). The problem is related to the fact that the `ngOnDestroy` was not called before for such `ControlContainer` and now it's being invoked.

I tried to put together a test case using TestBed, but it wasn't successful so far. The error that I'm getting while running tests in Google's codebase is (a part of the message):

```
  TypeError: Cannot read property 'registerControl' of null
      at packages/forms/src/directives/ng_form.ts?l=189
      at ZoneDelegate.invoke (packages/zone.js/lib/zone.ts?l=1148)
     ...
```

and that seems to happen when the input is changed (try clicking the button on the Stackblitz page I mentioned above).

If you see some things that might be causing this error - please let me know. Otherwise I will continue investigation and try to create a TestBed test.

Thank you.

llorenspujol commented:
Thank you @AndrewKushnir for the reproduction, I would look into it during this week and I will provide some feedback.

llorenspujol committed:
fix(core): call ngOnDestroy on useFactory providers that implement it and are provided at component level (#28738)

Fix ngOnDestroy to be called on useFactory providers that implement it
and are provided at component level. Fix applies to both ViewEngine and

llorenspujol commented:
I've finally got some time and **I have been able to reproduce it in a test**. I've pushed new changes with it. The test is crashing now, since the solution for it should be discussed, as it is non trivial.

About the [test](https://github.com/angular/angular/pull/28737/files#diff-9d49b87bd9f813d36f5da177bfee5465c78e2e266129c989bd2035307f364e9dR2647) and the reproduction, I just put 2 nested-form components and  forced to destroy one of those in order to call the `ngOnDestroy` of the prodived `ngModelGroup`, after that, I just change field on the 2nd `NestedForm` and the error occurs, since provided `ngModelGroup` has been technically destroyed (it's `ngOnDestroy` has been called).

What happens is that we are calling `ngOnDestroy` on a `ngModelGroup` provided by a factory function on that has not been instantiated on the component that we are destroying. So, it has been provided on that component, but is has not been instantiated on it. It has been instantiated on the context of it's parent.

So the question here is, should we be calling `ngOnDestroy` on providers provided by factory function that have non been instantiated on that component context? I would say no at first, but destroying factory providers provided on component level only if those providers have been instantiated on that component context may be a more complex functionality (both for Angular users and on Angular code), but it seems more natural right now on my first impressions.

Let me know your thoughts about it @AndrewKushnir .

AndrewKushnir commented:
Hi @llorenspujol,

Thanks a lot for additional investigation and putting together a test that reproduces the issue 👍 

I think we should try to keep the logic simple. So if a provider is defined on a given component and that component is destroyed, all providers should be destroyed as well. This logic is consistent with other provider types (except `useExisting` ones where it's clear that existing provider instance is used, thus it should not be destroyed). However we should perform some additional investigation to better understand the impact of this change to make sure there are no major use-cases (that a lot of people rely on) that become broken. So I can think of the following next steps that we can take:

- Perform an investigation related to the discussed use-case (perform Google search and look for cases/scenarios that might be affected).
- Analyze issues in Google's codebase to see how hard it'd be to update the code (to avoid relying on the old behavior) and if there are other usage patterns that might be affected. I can help with that.
- Create a test that would have a component that defines a provider using `useExisting` (where provider is originally defined in a different view) and see how the code behaves in this case (when that component is destroyed).

Thank you.

llorenspujol commented:
Hi @AndrewKushnir ,

About what you said: **_"if a provider is defined on a given component and that component is destroyed, all providers should be destroyed as well"_**, I understand that we should destroy providers provided by factory functions on a component even those providers are not instantiated and that destroyed component.

Although I understand that the logic is more simple, and this is good, I have some concerns about it. The fact that factory functions can provide a provider reference from some _'parent context'_ and destroy it leaves the door open to Angular users to reproduce what is done in the forms test case [example](https://github.com/angular/angular/pull/28737/files#diff-9d49b87bd9f813d36f5da177bfee5465c78e2e266129c989bd2035307f364e9dR2647). So even we refactor that code to use `useExisting`, the possibility to make a bad use of it would still be there.

What I would do is **one** of those 2 things:

1. Don't allow angular users to provide a provider reference from a 'parent' context using `useFactory`, so force them use `useExisting`.
2. Destroy providers provided by `useFactory` on the context they are created, not provided (prefered).

Said that, and in case that we finally decide to just call `ngOnDestroy` in all provided providers even they are not created on that context, I think we can refactor the code using `useExisting` in almost all the cases. The test example that is breaking now works as intended using `useExisting`. Although on that example we cover the functionality as expected, it can vary on some specific use cases since `useExisting` and `useFactory + skipSelf` don't have the same logical result. If you have more examples I may have enough information to find if there is something missing on this possible refactor.

Let me know your thoughts about it!

Thanks.

mhevery commented:
# Summary 

(I have discussed this with @AndrewKushnir and leaving our thoughts here for prosperity)

1. It seems that calling `ngOnDestroy` on any injectable is the right thing to do and is consistent behavior across framework. This should include `useFactory` cases as well which this PR solves.
2. With `useFactory` case there is complication that `useFactory` can return the same instance on multiple invocation:
  - If developer chooses to return same instance, then it is developers responsibility to make sure that `ngOnDestroy` does something reasonable.
    - There is no easy way for framework to know what "reasonable" action should be if the same instance is returned multiple times.
    - There is no easy/performant way for framework to determine and track that the same instance was returned and somehow delay the `ngOnDestroy`
    - It would be hard to explain in our documentation the corner cases of returning same instance by `useFactory` and than describe rules under which the `ngOnDestroy` will run. 

We believe that this PR fixes a bug. The fact that there is code out there which relies on this bug is unfortunate, but this should be treated as a bug fix. We expect that this will have small impact because to see difference in behavior one has to 1) use `useFactory` 2) return the same instance on multiple occasions, 3) have `ngOnDestroy` method which invalidates the object for subsequent usage. We don't believe this is a situation which is common enough. 


For existing code the simplest way to refactor for the new correct behavior is to change the injectable to have a ref-counter. Every time an instance of injectable is handed out by `usFactory` the counter is incremented. When framework calls `ngOnDestroy` the counter is decremented. When count reaches zero, `ngOnDestroy` performs the cleanup.

llorenspujol commented:
Thanks for the summary, it makes sense.

So for the refactoring to be enabled, I have pushed new changes adding a `refCount` to `AbstractFormGroupDirective`. This way both `NgModelGroup` and `FormGroupName` useFactory usages can be refactored if needed.

I also have modified the forms test that was breaking to use this refactoring, causing it to work as expected.

Let me know what do you think about this changes,

Thanks!

mhevery commented:
> So for the refactoring to be enabled, I have pushed new changes adding a `refCount` to `AbstractFormGroupDirective`. This way both `NgModelGroup` and `FormGroupName` useFactory usages can be refactored if needed.

Right, so from `core` point of view. This PR is good. If there are other tests failing, than it is those tests which need to be fixed, most likely through the `refCount` trick.

JoostK commented:
@llorenspujol FYI there's currently an issue with CI, so you won't get the tests green at the moment.

JoostK commented on packages/forms/src/directives/abstract_form_group_directive.ts:
```diff
@@ -72,6 +84,22 @@ export class AbstractFormGroupDirective extends ControlContainer implements OnIn
     return this._parent ? this._parent.formDirective : null;
   }
 
+  /**
+   * @description
+   * Returns the current number of references to this directive.
+   */
+  get refCount() {
+    return this._refCount;
+  }
+
+  /**
+   * @description
+   * Adds a reference into the total reference count.
+   */
+  addRef() {
+    this._refCount++;
+  }
+
```
Is there are reason for these being public? It seems like an implementation detail to me.

llorenspujol replied:
`refCount()` is an implementation detail but I set it public for consuming it in tests, we could maybe put an underscore to note that it is in fact private.

`addRef()` Needs to be public AFAIK since we need to be able to execute it when we provide the same instance multiple times in order to correctly do the cleanup.

llorenspujol replied:
My preferred solution would be just having a variable `_refCount`, for example to avoid having those methods on the 'public api', since they are in fact a hacky way to solve one specific problem.

But I was totally unable to run the build using only that variable, don't know why `NgModelGroup._refCount` was failing... (the error was _refCount is not a property of NgModelGroup). So I 'imagined' there would be some restriction or some step I don't know and I decided to put those methods to make it work.

AndrewKushnir replied:
I believe we should revert these changes in Forms based on https://github.com/angular/angular/pull/28737#issuecomment-728266633. Current understanding is that the problem with re-providing the same instance via `useFactory` should be taken care of on the apps (not framework) side. However I'd propose to keep a test for now (and mark it with `xit` vs `it`) to have more context and we can update or remove the test later.

The next step for this PR is to perform further investigation in Google's codebase (Angular team will handle that) to see if it's feasible to update the code to fix the issue with `ngOnDestroy`. We'll have more details after investigation and will make a final decision based on this data.

llorenspujol replied:
I may miss something but as far as I know without modifying `NgModelGroup` or `AbstractFormGroupDirective` with the capability to count it's references, and execute `ngOnDestroy` only when there is no reference left is the only way to refactor this use case. It still would have to be handled by the apps that consume it, executing `ngModelGroup.addRef()` whenever the same instance is provided.
Custom providers from apps could handle this problem the way the want, but in the case of NgModelGroup, apps have no way to prevent calling `ngOnDestroy` on it, as far as I know.

Other option that comes to mind, without touching framework code, would be making some runtime hack on the `__proto__ `of `NgModelGroup`, preventing this way the call of `ngOnDestroy`, but seems slow and too much complicated.

If there is another way to solve this bug (that I don't see) using refCounting and without touching the framework code, I would like to know, since now only this one's comes to my mind.

Thanks!

AndrewKushnir replied:
@llorenspujol we'd like to see whether we can refactor all affected usages in Google's codebase to avoid this pattern altogether (so we don't need to instrument the code with new API or monkey-patch field onto instances). We can reconsider the approach after initial investigation.

AndrewKushnir commented:
@llorenspujol sorry for the delay. We've discussed this question with the team once again and I'm going to close this PR for now, since landing this change would result in a lot of breakages. We'll keep the original ticket open and look into other possible solutions if we decide to perform some bigger updates to the DI system/API (in which case we can incorporate this behavior). Thank you.
